### ISSUE TITLE
Support deep objects for query parameters with deepObject style

### ISSUE THREAD

**Issue by bajtos** [OAI/OpenAPI-Specification#1706]
**Background**

Many applications expect deeply nested objects in input parameters, see the discussion in swagger-ui starting from this comment: https://github.com/swagger-api/swagger-ui/issues/4064#issuecomment-357417513 In [LoopBack](https://loopback.io), we are running into this problem too, see https://github.com/strongloop/loopback-next/pull/1679.

Consider a `filter` parameter defined as follows:

```yaml
parameters:
 filterParam:
   in: query
   name: filter
   schema:
     type: object
   style: deepObject
   explode: true
   description: Options for filtering the results
   required: false
```

Let's say the user wants to provide the following value, for example by entering a JSON into the text area rendered by swagger-ui:

```json
{
  "name": "ivan",
  "birth-date": {
    "gte": "1970-01-01"
  }
}
```

At the moment, the OpenAPI Specification v3 does not describe how to encode such value in a query string. As a result, OpenAPI clients like swagger-ui don't know how to handle this input - see the discussion in https://github.com/swagger-api/swagger-js/issues/1385 for an example.

**Proposal**

The following query string should be created by swagger-js client for the input value shown above.

```
filter[name]=ivan&filter[birth-date][qte]=1970-01-01
```

The proposed serialization style is supported by https://www.npmjs.com/package/qs, which is used by http://expressjs.com as the default query parser, which means that a vast number of Node.js API servers are already expecting this serialization style.

I am not sure if there is any existing formal specification of this format, I am happy to look into that once my proposal gets considered as acceptable in principle.

**Additional information**

Existing issues that are touching a similar topic:
 - style: form + object looses the parameter name? OAI/OpenAPI-Specification#1006
 - Are indexes in the query parameter array representable? OAI/OpenAPI-Specification#1501

Two older comments from swagger-js that may be relevant:

https://github.com/swagger-api/swagger-js/pull/1140
> Limitations:
> deepObject does not handle nested objects. The specification and swagger.io documentation does not provide an example for serializing deep objects. Flat objects will be serialized into the deepObject style just fine.

https://github.com/swagger-api/swagger-js/pull/1140#issuecomment-333017825
> As for deepObject and nested objects - that was explicitly left out of the spec, and it's ok to just Not Support It™.

**Comment by earth2marsh** [comment]
Are there any other examples of these nested deepObjects outside of Express? The more widespread a pattern, the more likely it is to be considered. Myself, I have an aversion to passing such a complicated object in the query string. Any insight into why Express even landed on this pattern?

**Comment by darrelmiller** [comment]
We were a bit stuck when allowing the `deepObject` style.  We understood that users wanted this capability but there is no standard definition of what that serialization format looks like.  We had a few choices, allow it and define our own standard and hope implementations followed it.  Don't allow it because there are no standards, or add it in, say nothing about its format and hope that a default serialization format emerges.

If we can get some confidence that the `qs` package is becoming a defacto standard and we can create an accurate description of the serialization, then I have no issue recommending that we include that description in a future minor release of the spec.

**Comment by rmunix** [comment]
What about the URL max length limits? I think one of the reasons why a standard for object serialization in the URL is hard to materialize is because of the URL max length problem, the URL just wasn't intended to pass data in this way.  Depending on the browser and/or server software being used the URL max length varies but in general it is small when compared to how much data can be transmitted with other methods like POST or PUT.  It certainly would work for small objects but people tend to inadvertently abuse these kinds of APIs by passing bigger than allowed payloads.

**Comment by handrews** [comment]
@rmunix regarding URL max length problems, I was very happy to see the HTTP SEARCH method draft RFC revived last month: https://tools.ietf.org/html/draft-snell-search-method-01

**Comment by louisl** [comment]
> Are there any other examples of these nested deepObjects outside of Express? The more widespread a pattern, the more likely it is to be considered. Myself, I have an aversion to passing such a complicated object in the query string. Any insight into why Express even landed on this pattern?

http://esbenp.github.io/2016/04/15/modern-rest-api-laravel-part-2/

I use a modified version of https://github.com/esbenp/bruno referenced in the article above in a few apis, it's extremely useful for including related data and search filtering. I'm not really sure how to define those in a spec. I do appreciate that the url strings could get silly long to the point of failing if abused, but without this sort of thing searches and such would have to be actioned as POST requests or limited to basic GET request params.  It seems to me no ones really come up with a holy grail API solution for complex search queries so it's a bit of a free for all at the moment.

**Comment by bajtos** [comment]
**Ruby on Rails**

Seems to use the same approach. 

Reference: https://edgeapi.rubyonrails.org/classes/Hash.html#method-i-to_query
Source code: https://github.com/rails/rails/blob/b5302d5a820b078b6488104dd695a679e5a49623/activesupport/lib/active_support/core_ext/object/to_query.rb#L61-L86

Example code:

```ruby
require "activesupport"

data = {
  "name" => "David",
  "nationality" => "Danish",
  "address" => {
    "street" => "12 High Street",
    "city" => "London",
  },
  "location" => [10, 20],
}
print data.to_query("person")
```

Produces the following query string, I have urldecoded and reformatted it for better readability:

```
person[address][city]=London&
person[address][street]=12+High+Street&
person[location][]=10&
person[location][]=20&
person[name]=David&
person[nationality]=Danish
```

Notice that array items are using an empty index, i.e. `person[location][]=10`, instead of `person[location][0]=10`.

**Comment by louisl** [comment]
Not API specific, but jQuery can generate nested array url params from objects.

http://api.jquery.com/jquery.param/

**Comment by benhaynes** [comment]
We have been trying to use OpenAPI 3.0 for a while now, but the lack of support for nested deepObjects has kept us from using this spec. We have a dynamic API that allows for relatively complex filtering, for example: `filter[<field-name>][<operator>]=<value>`

```
GET /items/users?filter[category][eq]=vip
```

[Our API Reference for this filtering](https://docs.directus.io/api/reference.html#filtering)

Is there any hope for this being supported in the future or should we "move on"?

**Comment by ewrayjohnson** [comment]
@bajtos: On October 11, 2018 you wrote "I'll try to find some time to fix swagger-js in the next few weeks." Where do you stand on this?

**Comment by bajtos** [comment]
Eh, I didn't even started 😢 Feel free to contribute the fix yourself.

**Comment by darrelmiller** [comment]
@benhaynes Sorry if this comes across as snarky, it's not intended to, it's just I'm in a rush and I don't know how to ask this in a friendly/sincere way.  

What would you like us to do?  Pick a winner from the many options?  Design our own format?  If we pick a format that is incompatible with what you currently use, would you switch?  Should we find a way of supporting many different formats?

**Comment by benhaynes** [comment]
Hey @darrelmiller — not snarky at all, I sincerely appreciate the response as it maintains some momentum in the discussion!

We're certainly not trying to _force_ the spec to follow our format, and understand your position of not wanting to blaze ahead without a standard to follow. To answer your question honestly, if the option you choose is incompatible with our method, then we wouldn't be able to use OpenAPI since we can't introduce a breaking change into our platform's filtering. Still, we'd support your team's decision if they think a different direction is a better solution.

I'm not sure how "extensible" your spec/codebase is, but support for multiple (even optional) serialization formats seems the most promising. Perhaps leaving these as _unofficial_ until a "winner" eventually surfaces. In our experience, industry-leading frameworks offering solutions is the most efficient way for a de facto standard to emerge.

Our proposal is to support a deeply nested structure, such as: `param[nest1][nest2]=value`, where there can be `1,2,3,n` levels of nesting. The comments here might be biased, but it seems that most are either already using this pattern or are recommending it.

Thanks again. I'd love to hear your (or anyone else's) thoughts on this approach!

**Comment by WellingGuzman** [comment]
Hey @darrelmiller, I would like to understand the OpenAPI position on this. Is the reason to not support the outcome `/items/users?filter[category][eq]=vip` because it's not a standard or because the parameter definition format is not part of the standard?

Also, in the technical side, will this bring a complex/breaking change and it requires much more time?

In term of where else this format is supported I would like to add another example.

PHP

Example:

```php
<?php

$data = [
  'person' => [
    'name' => 'David',
    'nationality' => 'Danish',
    'address' => [
      'street' => '12 High Street',
      'city' => 'London',
    ],
    'location' => [10,20],
  ]
];

echo http_build_query($data);
```

Result:

```
person[name]=David&
person[nationality]=Danish
&person[address][street]=12+High+Street
&person[address][city]=London
&person[location][0]=10
&person[location][1]=20
```

The result is urldecoded.

Also PHP automatically parse these parameters into an array. Passing the result above into query string  will result in the original array.

**Comment by apimon** [comment]
Arrrgh, hit that wall, too.... great example of a not so obvious limitation that might kill your whole project dev setup and workflow.

**Comment by okirmis** [comment]
Ran into the same problem (using Rails), also with a parameter allowing for dynamic filtering.

The funny thing is, that swagger-editor does generate deeper nested objects as placeholder text for the parameters text area when clicking "try it out":
```
      - in: query
        name: filter
        schema:
          type: object
          properties:
            boolean_param:
              type: boolean
              nullable: true
            some_enum_of_types:
              type: array
              items:
                type: string
              nullable: true
          default: {}
        style: deepObject
```
results in
```
{
  "boolean_param": true,
  "some_enum_of_types": [
    "string"
  ]
}
```
which will be silently ignored. I know that this is actually a bug in swagger-editor, but it shows that it would be more consistent to allow deeper nested objects.

**Comment by rijkvanzanten** [comment]
What needs to be done to move on with this @earth2marsh? You asked for a couple examples outside of Express, which I think have been provided in the messages above. I can help out writing some of the needed documents for this in a PR if that helps.

**Comment by darrelmiller** [comment]
Does someone have a proposal of how we can fix the OpenAPI specification to allow it to tell tooling which flavour of deepNesting object serialization to use?  

If someone wants to tackle this, here are some things to consider when creating the proposal:
- Which serialization flavors exist?  Will they be enumerated in the spec?  
- Will it be required that all tooling support all flavours?  
- If not, what should tooling do when it hits a flavour it doesn't support?
- Do we have commitment from tooling implementors to implement this feature?

**Comment by rijkvanzanten** [comment]
> Which serialization flavors exist? Will they be enumerated in the spec?

I hope I'm understanding what you're asking correctly, but based on the comments above

input

```
parameters:
 filterParam:
   in: query
   name: filter
   schema:
     type: object
   style: deepObject
   explode: true
   description: Options for filtering the results
   required: false
```

```
{
  "name": "ivan",
  "birthdate": {
    "gte": "1970-01-01"
  }
}
```

should result in

output

```
filter[name]=ivan&filter[birthdate][qte]=1970-01-01
```

The main difference seems to be in how nested arrays are handled. There seem to be slight differences between [PHP](https://github.com/OAI/OpenAPI-Specification/issues/1706#issuecomment-458949282), Node and [Ruby on Rails](https://github.com/OAI/OpenAPI-Specification/issues/1706#issuecomment-439421031). 

Both PHP and [qs*](https://npmjs.com/packages/qs) seem to favor the `&numbers[0]=5&numbers[1]=10` syntax

```php
$params = array(
  'filter' => array(
      'numbers' => array(5, 10)
  )
);

echo http_build_query($params);
// => filter%5Bnumbers%5D%5B0%5D=5&filter%5Bnumbers%5D%5B1%5D=10
```

```js
var qs = require('qs');

var result = qs.stringify({
  filter: {
    numbers: [0, 10]
  }
});

console.log(result);
// => filter%5Bnumbers%5D%5B0%5D=0&filter%5Bnumbers%5D%5B1%5D=10
```

In normal non-nested use the Node built-in `querystring` and Swagger favor `&numbers=5&numbers=10`:

```js
var querystring = require("querystring");

var result = querystring.encode({
  numbers: [5, 10]
});

console.log(result);
// => numbers=5&numbers=10
```

Combined by the comment @bajtos' comment about Ruby's implementation, this leaves us in a scenario where there are three different outputs for this given input for `filter`:

```json
{
  "name": "ivan",
  "birthdate": {
    "gte": "1970-01-01"
  },
  "numbers": [5, 10]
}
```

```
&filter[name]=ivan
&filter[birthdate][gte]=1970-01-01
&filter[numbers]=5
&filter[numbers]=10
```

```
&filter[name]=ivan
&filter[birthdate][gte]=1970-01-01
&filter[numbers][0]=5
&filter[numbers][1]=10
```

```
&filter[name]=ivan
&filter[birthdate][gte]=1970-01-01
&filter[numbers][]=5
&filter[numbers][]=10
```

Seeing that different languages and implementations are already out there, I don't think we can pick one over the other. (Eg PHP's `parse_str` doesn't work with 1, while Ruby's won't parse 2).

`qs` seems to support all three of these flavors in parsing.

Re: your comment Oct 29 2018

> If we can get some confidence that the `qs` package is becoming a defacto standard and we can create an accurate description of the serialization, then I have no issue recommending that we include that description in a future minor release of the spec.

Seeing the ~22 million weekly downloads, I think it's safe to say that `qs` has become the defacto query string parser in Node JS. 

I don't know enough about the governance of the project nor the extended OpenAPI family of projects to make any comments about the other questions mentioned.

**Comment by philleepflorence** [comment]
I concur and second, @rijkvanzanten...
```
&filter[name]=ivan
&filter[birthdate][gte]=1970-01-01
&filter[numbers][]=5
&filter[numbers][]=10
```

This format is rather important! It would allow for an array of arrays (PHP) or an array of objects (Node).

**Comment by br3nt** [comment]
Couldn't query params be defined in the same way as body objects by specifying a schema?  I'll add my thought about serialization after the example.

So, as an example: 
```js
params = {
  some_query_object: {
    hello: "world",
    foo: [
      { value: "blerg" },
      { value: "blah" }
    ],
    bar: [1, 2, 3]
  }
}

jQuery.param(params)
```
Serializes to this query string:
```
 some_query_object[hello]=world
&some_query_object[foo][0][value]=blerg
&some_query_object[foo][1][value]=blah
&some_query_object[bar][]=1
&some_query_object[bar][]=2
&some_query_object[bar][]=3
```

And, I would expect to be able to write something like this: 
```yaml
paths:
  "/api/example":
    post:
      produces:
        - application/json
      parameters:
        - name: some_query_object
          in: query
          description: The object in the query string
          required: false
          schema:
            "$ref": "#/definitions/some_query_object"
definitions:
  some_query_object:
    type: object
    properties:
      hello:
        type: string
      foo:
        type: array
        items:
          "$ref": "#/definitions/nested_object"
      bar:
        type: array
        items:
          type: string
  nested_object:
    type: object
    properties:
      value:
        type: string
```

In regards to query string serialization, the only case OpenAPI really needs to account for is the various array assignment types (The python example above is clearly `type: string` so can be ignored).

I would suggest adding a single option `arrayAssignmentType`.
It only applies to arrays.  Not objects/maps.
I'm not sure if it should be a global option or per property.  Global should be fine as its generally enforced by frameworks/libraries.

`arrayAssignmentType` would have the the following options:

`direct` (or `traditional` like in [jQuery param docs](https://api.jquery.com/jquery.param/)), where the values are assigned directly to the key:
```
item=a&item=b
OR
obj[item]=a&[item]=b
```

`accessor`, where values are assigned via an `[]` or `[int]` accessor:
```
item[]=a&item[]=a
OR
obj[item][1]=a&item[2]=b
OR
obj[item][]=a&obj[item][]=b
OR
obj[item][1]=a&[item][2]=b
OR
obj[][value]=a&obj[][value]=b
OR
obj[1][value]=a&obj[2][value]=b
```

The naming can be worked out later.

In regards to the `accessor` option, it would be interesting to see if various frameworks support both formats (`[]`  vs `[1]`).  I would expect the answer is both based on how JQuery serializes params:
```js
$.param({ a: [1, 2, 3] })
"item[]=1&item[]=2&item[]=3 "

$.param({ item: [{ x: 'a' }, { x: 'b'}, { x: 'c'}] })
"item[0][x]=a&item[1][x]=b&item[2][x]=c"
```

I think the default value should be `arrayAssignmentType: accessor` just because that format is more explicit in how arrays are defined (...also "to accommodate modern scripting languages and frameworks such as PHP and Ruby on Rails" — jQuery docs lols)

**Comment by br3nt** [comment]
Ok, I just tested how Rails handle the different array accessors in the query string:

```
?item=1&item=2&item=3
# { 'item' => '3' }

?item[]=1&item[]=2&item[]=3
# { 'item' => ['1', '2', '3'] }

?item[1]=1&item[2]=2&item[3]=3
# { 'item' => { '1' => '1', '2' => '2', '3' => '3' } }
```

So, @darrelmiller, following on from my previous comment, perhaps there can be 4 options for `arrayAssignmentType`... `direct` (or `traditional`), `emptyAccessor`, `numericAccessor`, `dynamicAccessor`.  Again, the naming is up for debate, it's more about the concept.

Likewise, perhaps `arrayAssignmentType` should be both a global option and a per-property option, to account for all scenarios.

**Comment by bajtos** [comment]
Thank you all for pushing this discussion forward ❤️ 

@br3nt's proposal seems very reasonable to me.

>  perhaps there can be 4 options for arrayAssignmentType... direct (or traditional), `emptyAccessor`, `numericAccessor`, `dynamicAccessor`. Again, the naming is up for debate, it's more about the concept.

It's not clear to me what is the difference between `numericAccessor` and `dynamicAccessor`, but that's a minor detail that can be figured out while working on the actual spec proposal.

> perhaps `arrayAssignmentType` should be both a global option and a per-property option, to account for all scenarios.

+1

Do we want this option to be per-property (i.e. inside a schema definition) or per-parameter (i.e. inside an operation parameter definition)?

I prefer the later.

If we allow `arrayAssignmentType` on per-property basis, then we can end up with different properties using different array assignment types. That feels too complicated to me.

```
?q[hello]=world
# arrayAssignmentType: numericAccessor
&q[foo][0]=blerg
&q[foo][1]=blah
# arrayAssignmentType: emptyAccessor
&q[bar][]=1
&q[bar][]=2
&q[bar][]=3
```

Anyhow, if there is a need for this extra complexity, then perhaps we can allow `arrayAssignmentType` at three levels?
- global
- specific to an operation parameter
- specific to a property in a schema

The part important for me is that property-level or parameter-level setting should overwrite any globally-defined setting.

**Comment by darrelmiller** [comment]
Thanks everyone for the input here.  This is exactly the type of information we needed, but didn't have when were originally defining the deepObject style.

I am a little concerned about the potential complexity for tooling to support a range of different formats, especially now we are talking about defining it at different levels and allowing overloading.  I wonder if this is unnecessary complexity.

There are two major use-cases for being able to define how an object is serialized in the query string.
1) In client tooling where it is necessary to construct a URL so an API can be called.
2) In server side tooling to parse a URL and reconstruct a parameter value as an object.

If we go down the path of creating configuration parameters that reflect how a particular library/language/platform does the serialization then we are leaking implementation details through the OpenAPI description.  This is somewhat defeating what OpenAPI is trying to do. 

There doesn't seem to be a big a variation in formats as I once believed and I wonder if there is value in OpenAPI just picking the most compatible format and completely specifying it.  From what I can tell it would be possible for us to add wording to the specification that would allow us to say `deepObject` follows these rules:
A parameter named `q` that has a value that looks like this,
```
{
    hello: "world",
    foo: [
      { value: "blerg" },
      { value: "blah" }
    ],
    bar: [1, 2, 3]
  }
```
would serialize as

```
? q[hello]=world&q[foo][0][value]=blerg&q[foo][1][value]=blah&q[bar][0]=1&q[bar][1]=2&q[bar][2]=3
```
This assumes that all numeric keys are references to elements of an array and all textual keys are references to elements of a map.  The keys could be optionally quoted to allow keys for maps to be numbers.

The use of the empty indexer seems problematic to me because URL query parameters usually do not have an order that is semantically significant but array values usually are ordered.  

If we assume most frameworks support both the numeric accessor and the empty accessor, then limiting `deepObject` to just using the numeric accessor should provide reasonable compatibility.  It does appear this would be a problem for Rails.  Some kind of OpenAPI specific parser would be needed to support the deserialization process.  However, whichever format we pick, one or more platforms are going to have to do work to support it.

I do think that having a few platforms need to build libraries to conform to a consistent approach is a better long term outcome than creating a parameterization system that effectively describes the platform.  If we did that we might as well create styles for `rubyDeepObject`, `phpDeepObject`, `nodeDeepObject`

I do however recognize there is a need to retroactively describe an API that has selected a non-standard serialization format.  There is a solution to that.  Instead of using the `style` parameter which is designed to be quick an easy way to identify a common serialization style, you can use content and a media type.

```
parameters:
  - name: q
    content:
      application/vnd.ruby.deepObject: {}
```

The purpose of media types is to describe semantics, syntax and serialization of data going over the wire.  Normally, they are used for request and response bodies.  However, OpenAPI allows defining a content object for parameters also.  This also allows defining a schema if it is necessary to identify if a particular property is an object or an array when the syntax in the URL does not disambiguate.

In the past I was not a fan of OpenAPI picking a winner when it comes to the serialization format.  However, considering the impact of the alternative approach, I have to wonder if it is the right thing to do.   How would people feel if OpenAPI explicitly described a format for `deepObject` and required using `content` and a custom media type to describe parameters that don't match that standard?

**Comment by everlast240** [comment]
Hey guys,
I've been ready & trying different things for some time now, but can't get the standard JSONAPI sparse fieldsets params to work with swagger-ui... 
JSONAPI specifies a way to fetch only required attributes of the included resources. E.g.:
```json
fields[resource1]=attrib1,attrib2,attrib3&fields[resource2]=attrib1
// example:
fields[person]=first_name,last_name,age&fields[address]=city,full_address,latitude,longitude
```
This all works for my endpoints, but no matter how I format the payload I enter in SwaggerUI and/or setup the swagger JSON, I can't get it to work...
For example, using the endpoint(s) works like that:
```
curl -X GET "http://localhost:5000/person/39?include=address,address.country&fields[person]=first_name,last_name,age&fields[address]=city,full_address,latitude,longitude"
```
Tried different swagger JSON setups (e.g.with deepObject etc), and payloads in the swagger-ui text field (e.g.: `{person: "first_name,last_name,age", address: "city,full_address,latitude,longitude"}`), but no success for now...
Part of the complication is that the object / hash keys (model / resource name) are dynamic: they depend on the included resources (in the compound document) and I can't statically describe the schema definition.

Any ideas? Help is appreciated!
PS: Using Grape (https://github.com/ruby-grape/grape/) and the Netfix fast_jsonapi (https://github.com/Netflix/fast_jsonapi).

**Comment by br3nt** [comment]
@darrelmiller, picking one format rather than a more versatile solution will do more harm than good.

It will alienate the frameworks that don't support whatever format is chosen.  Using the example of numeric array accessors...

ASP.NET apps would be excluded as they use what I described as `direct` or `traditional` style array assignment.  ASP.NETs qs format is more similar to:
```
?q.hello=world&q.foo.0.value.=blerg&q.foo.1.value=blah&q.bar=1&q.bar=2&q.bar=3
```

JS frameworks which use the [qs](https://www.npmjs.com/package/qs) library, like ExpressJS and LoopBack, will also be excluded.  As per the `qs` documents:
> Any array members with an index of greater than 20 will instead be converted to an object with the index as the key. This is needed to handle cases when someone sent, for example, a[999999999] and it will take significant time to iterate over this huge array.

E.g:
```
?foo[21]=bar
{"foo":{"21":"bar"}}
```
However, `qs` supports both the `direct` or `traditional` style and the `emptyAccessor`.  E.g:
```
?foo=bar&foo=baz
{"foo":["bar","baz"]}

?foo[]=bar&foo[]=baz
{"foo":["bar","baz"]}
```

Would tooling implements be expected to parse `foo[21]` as an array or a map? Is that fair?  Why stop at 20?  It's kind of arbitrary.  At least the empty array accessor `[]` is consistent.  It's always an array which gets populated based on the order of the query string.

PHP apps would be fine as PHP arrays work basically like hashes/maps.

Picking one option over another tells frameworks that their decisions are right or wrong. I don't think that's the job of OpenAPI Specifications.  A better place to decide on standardization would be in a W3C proposal where the community could come to a consensus together.

Your concern about additional complexity is fair enough.  However, I think using only content media types sweeps the problem under the rug a little.  Small frameworks would be disadvantaged by this decision.

I think it reasonable to expect some minimum support from OpenAPI Specification for the array accessor styles of `foo[]=a`, `foo[0]=a` and `foo=a&foo=b` and it's not unreasonable to expect tooling implementers to support these three styles.

I think it would also be reasonable to assume that most (if not all) frameworks would offer consistent array accessor parsing functionality.  As such a global option specifying array accessor styles would be sufficient to tell tooling which to use (rather than on a per-property or per-path basis which I agree is way too complex).

I don't think the above could be considered as leaking framework functionality into the OpenAPI-Specification.

Anything more complex than the above, then I also agree that content media types would be a good solution.

**Comment by ewrayjohnson** [comment]
My apologies for not responding for while.  This may seem to be to simplistic considering the complexity of other comments, but what if content (each qs param value)  be subject to JSON.parse/JSON.stringify subject to URL/URI decoding/encoding as needed with whitespace removal.

For example:
{
  "name": "ivan",
  "birthdate": {
    "gte": "1970-01-01"
  },
  "numbers": [5, 10]
}

can have whitespace removed as

{"name":"ivan","birthdate":{"gte":"1970-01-01"},"numbers":[5,10]}

encoded as

?param1=%7B%22name%22%3A%22ivan%22%2C%22birthdate%22%3A%7B%22gte%22%3A%221970-01-01%22%7D%2C%22numbers%22%3A%5B5%2C10%5D%7D

It may be possible that only certain characters need to be encoded (e.g. ampersand as %26)

**Comment by louisl** [comment]
> param1=%7B%22name%22%3A%22ivan%22%2C%22birthdate%22%3A%7B%22gte%22%3A%221970-01-01%22%7D%2C%22numbers%22%3A%5B5%2C10%5D%7D

Thats pretty much the same as defining param1 as a string then decoding and objectifying the string which is how I'm getting around the issue for now.

The problem with that approach is there's no way in the spec to tell what a valid object/array would be for param1 or that param1 should even be an object/array.

**Comment by bajtos** [comment]
> `param1=%7B%22name%22%3A%22ivan%22%2C%22birthdate%22%3A%7B%22gte%22%3A%221970-01-01%22%7D%2C%22numbers%22%3A%5B5%2C10%5D%7D`

In LoopBack, we support both style. When the parameter is described as array or object with `style: deepObject`, then:
- First, we let `qs` to parse `[]` notation into an object or an array. If this returns an object value then we are done.
- Otherwise if the query provides a string value, then we deserialize the value from JSON. 

See https://github.com/strongloop/loopback-next/blob/d1eef59928fe8afb56e8bfce3d04e1e3a656ae2b/packages/rest/src/coercion/coerce-parameter.ts#L169-L200

> The problem with that approach is there's no way in the spec to tell what a valid object/array would be for param1 or that param1 should even be an object/array.

Yes, the spec does not provide any means how to express this style. However, I can imagine adding e.g. a new `style` variant to describe JSON-serialized values.

For example:

```yaml
   in: query
   name: filter
   schema:
     type: object
   style: json
   description: Options for filtering the results
   required: false
```

**Comment by louisl** [comment]
> Yes, the spec does not provide any means how to express this style. However, I can imagine adding e.g. a new `style` variant to describe JSON-serialized values.
> 
> For example:
> 
> ```yaml
>    in: query
>    name: filter
>    schema:
>      type: object
>    style: json
>    description: Options for filtering the results
>    required: false
> ```
Something like that would be useful, I wonder if it could be expanded to somehow use a schema for validity of the JSON object?

**Comment by hkosova** [comment]
@ewrayjohnson 

> what if content (each qs param value) be subject to JSON.parse/JSON.stringify subject to URL/URI decoding/encoding as needed with whitespace removal.
> For example:
> ```json
> {
>   "name": "ivan",
>   "birthdate": {
>     "gte": "1970-01-01"
>   },
>   "numbers": [5, 10]
> }
> ```
>
> can have whitespace removed as
> ```json
> {"name":"ivan","birthdate":{"gte":"1970-01-01"},"numbers":[5,10]}
>```
> encoded as
> ```
>?param1=%7B%22name%22%3A%22ivan%22%2C%22birthdate%22%3A%7B%22gte%22%3A%221970-01-01%22%7D%2C%22numbers%22%3A%5B5%2C10%5D%7D
> ```

I believe this particular use case is already covered by the [`content`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#parameterContent) keyword instead of `schema`/`style`/`explode`.

```yaml
parameters:
  - in: query
    name: param1
    content:
      application/json:
        schema:
          type: object
          properties:
            name:
              type: string
              example: ivan
          ...
```

**Comment by chargio** [comment]
This would be using this feature extensively. We have many use cases where the user needs this feature

**Comment by kurko** [comment]
To be able to document things, I've been using the following, which is not ideal:

```yaml
parameters:
  - schema:
    type: string
  in: query
  name: 'filter[search][eq]'
```

This is a particular problem with standards like jsonapi.org.

**Comment by ahx** [comment]
@kurko What problems did you have with this flat-nested-syntax (`name : 'filter[search][eq]`') ?
I am doing the same. The only problem I have had with this so far was that we cannot use `anyOf` when using this flat naming scheme compared when using a real `object`. Other than that, I would prefer using that syntax, because it is shorter.

**Comment by kurko** [comment]
@ahx you won't be able to use tools to automatically validate your requests, unless they cast the string into some sort of dictionary/hash/object. For instance, with Ruby (I'm trying out Committee gem for request validations), that spec would be converted to `{ "filter[search][eq]" => "value" }` (note the string as key) whereas in fact it would need to be:

```rb
{
  filter: {
    search: {
      eq: "value"
    }
  }
}
```

I have to check whether none of the tools I'm using isn't trying to hack this and convert to an object. Haven't got to that point yet in development.

**Comment by ahx** [comment]
@kurko Hey, I am also using Ruby 🚀 This is a known issue for committe, but this [was solved](https://github.com/ahx/openapi_first/issues/40) in `[openapi_first](https://github.com/ahx/openapi_first)` (which is an alternative to committee created by myself 👋 )

**Comment by PaulWay** [comment]
We needed to support filters like these in our API, so I've written this code:

https://github.com/PaulWay/django_filter_deep_object_param

It's kind of a copy and paste, and I'm still working on tests and better documentation.  But hopefully that can be a basis of some of the work to support deep object querying.  I'm also completely open to any suggestions on how to do this better :smile: 

It works for us particularly because PostgreSQL 11 and above support deep querying into JSON objects, but it also supports queries through Django relationships.

**Comment by ewrayjohnson** [comment]
> We needed to support filters like these in our API, so I've written this code:
> 
> 
> 
> https://github.com/PaulWay/django_filter_deep_object_param
> 
> 
> 
> It's kind of a copy and paste, and I'm still working on tests and better documentation.  But hopefully that can be a basis of some of the work to support deep object querying.  I'm also completely open to any suggestions on how to do this better :smile: 
> 
> 
> 
> It works for us particularly because PostgreSQL 11 and above support deep querying into JSON objects, but it also supports queries through Django relationships.
> 
> 

The double underscore delimiter is not foolproof when properties/relations contain leading or trailing underscores.  I suggest a character not valid in loopback property/relation names but valid in url query string preferably without encoding (possibly a tilde or believe it or not a period).  Caveat:  All this is just off the top of my head late at night with no research.

**Comment by PaulWay** [comment]
Good thinking @ewrayjohnson - good checks to do.

This cannot prevent possible exploration of the database schema using Django's relations, and there are security implications.  It all depends on the field you're querying and how you're expressing your relations in Django.

**Comment by UkonnRa** [comment]
In JSON:API, we would like do design QueryParam **filter** like [this](https://jsonapi.org/recommendations/#filtering) and https://github.com/json-api/json-api/issues/1445:

```
GET /authors?page[articles][offset]=10&page[articles][limit]=5 <-- Nested object

GET /comments?filter[post]=1,2&filter[author]=12 <-- Primitive array inside the object
```

So:

1. I think nested object is useful for more powerful API design
2. About the primitive array inside deep objects, as [this comment](https://github.com/swagger-api/swagger-ui/issues/6672#issuecomment-736311886) said, we should fine some way to let users customize the representation

**Comment by creckord** [comment]
Spring also supports deep objects using dot-notation, i.e. GET /authors?page.articles.offset=10&page.articles.limit=5

When going forward with this proposal, what would be a really great addition would be the option to skip the root parameter name. This would neatly solve #445:

```
paths:
  /pageable-endpoint:
     get:
         parameters:
            - name: query
               in: query
               schema:
                   type: string
            - name: page
               in: query
               required: false
               schema:
                   $ref: '#/components/schemas/Page'
               # option to skip 'page' in the effective query parameter name here
               style: deepObject
components:
   schemas:
      Page:
         type: object
         properties:
           page:
              type: integer
           items:
              type: integer
```
could result in

```
GET /pageable-endpoint?query=foo&page=5&items=100
```

instead of
```
GET /pageable-endpoint?query=foo&page.page=5&page.items=100
```

**Comment by hkosova** [comment]
@creckord 
> When going forward with this proposal, what would be a really great addition would be the option to skip the root parameter name. This would neatly solve (some aspects of) #445:
> ...
> could result in
> ```
> GET /pageable-endpoint?query=foo&page=5&items=100
> ```
This particular example is already supported by [`style: form` + `explode: true`](https://spec.openapis.org/oas/v3.1.0.html#style-examples) (which is the default style for query params) rather than `deepObject`.

**Comment by creckord** [comment]
@hkosova 
> This particular example is already supported by [`style: form` + `explode: true`](https://spec.openapis.org/oas/v3.1.0.html#style-examples) (which is the default style for query params) rather than `deepObject`.

Ah, good to know. Then it's just the Java+okhttp client generator that does not support this particular style.

What would happen with that style if the Page schema had nested object-type properties?

**Comment by hkosova** [comment]
@creckord 
> What would happen with that style if the Page schema had nested object-type properties?

Undefined behavior / not supported.

Nested structures should ideally be handled by `deepObject`, which is what this discussion is about.

**Comment by mrleblanc101** [comment]
Hi, 
On a project I'm working on, we are trying to encore these filters.
We tried `style: form, explode: true` and `style: deepObject, explode: true` without luck.
```
{
  persona: [
    'positionner_1er',
    'vendre',
    'selectionner',
    'maximiser',
    'positionner_2e'
  ],
  type: [ 'programme', 'parcours', 'activite', 'capsule' ],
  theme: [
    'ressources-humaines',
    'numerique',
    'marketing-com',
    'vente',
    'finance',
    'gestion',
    'innovation'
  ],
  difficulte: [ 'avance', 'debutant', 'intermediaire' ],
  region: [
    'outaouais',
    'mauricie',
    'bas-saint-laurent',
    'laurentides',
    'province-du-quebec',
    'abitibi-temiscamingue',
    'capitale-nationale',
    'montreal',
    'centre-du-quebec',
    'chaudiere-appalaches',
    'cote-nord',
    'estrie',
    'gaspesie-iles-de-la-madeleine',
    'lanaudiere',
    'laval',
    'monteregie',
    'nord-du-quebec'
  ],
  duree_intervalle: [ '*-3600', '3600-10800', '10800-28800', '28800-230400', '230400-*' ],
  dispo_dans_catalogue: true
}
```
We need an encoding like this:
```
filters[persona][]=positionner_1er&filters[persona][]=vendre&filters[persona][]=selectionner&filters[persona][]=maximiser&filters[persona][]=positionner_2e&filters[type][]=programme&filters[type][]=parcours&filters[type][]=activite&filters[type][]=capsule&filters[theme][]=ressources-humaines&filters[theme][]=numerique&filters[theme][]=marketing-com&filters[theme][]=vente&filters[theme][]=finance&filters[theme][]=gestion&filters[theme][]=innovation&filters[difficulte][]=avance&filters[difficulte][]=debutant&filters[difficulte][]=intermediaire&filters[region][]=outaouais&filters[region][]=mauricie&filters[region][]=bas-saint-laurent&filters[region][]=laurentides&filters[region][]=province-du-quebec&filters[region][]=abitibi-temiscamingue&filters[region][]=capitale-nationale&filters[region][]=centre-du-quebec&filters[region][]=chaudiere-appalaches&filters[region][]=cote-nord&filters[region][]=estrie&filters[region][]=gaspesie-iles-de-la-madeleine&filters[region][]=lanaudiere&filters[region][]=laval&filters[region][]=monteregie&filters[region][]=nord-du-quebec&filters[duree_intervalle][]=%2A-3600&filters[duree_intervalle][]=3600-10800&filters[duree_intervalle][]=10800-28800&filters[duree_intervalle][]=28800-230400&filters[duree_intervalle][]=230400-%2A
```
But all with `deepObject` our params end-up overwritting each other:
```
filter[persona]=positionner_1er&filter[persona]=vendre&filter[persona]=selectionner&filter[persona]=maximiser&filter[persona]=positionner_2e&filter[type]=programme&filter[type]=parcours&filter[type]=activite&filter[type]=capsule&filter[theme]=ressources-humaines&filter[theme]=numerique&filter[theme]=marketing-com&filter[theme]=vente&filter[theme]=finance&filter[theme]=gestion&filter[difficulte]=avance&filter[difficulte]=debutant&filter[difficulte]=intermediaire&filter[region]=outaouais&filter[region]=mauricie&filter[region]=bas-saint-laurent&filter[region]=laurentides&filter[region]=province-du-quebec&filter[region]=abitibi-temiscamingue&filter[region]=capitale-nationale&filter[region]=centre-du-quebec&filter[region]=chaudiere-appalaches&filter[region]=cote-nord&filter[region]=estrie&filter[region]=gaspesie-iles-de-la-madeleine&filter[region]=lanaudiere&filter[region]=laval&filter[region]=monteregie&filter[region]=nord-du-quebec&filter[duree_intervalle]=*-3600&filter[duree_intervalle]=3600-10800&filter[duree_intervalle]=10800-28800&filter[duree_intervalle]=28800-230400&filter[duree_intervalle]=230400-*&filter[dispo_dans_catalogue]=true&sort=created_at&append=pourcentage_complete
```
<img width="1225" alt="Capture d’écran, le 2021-07-06 à 12 16 19" src="https://user-images.githubusercontent.com/17392251/124634189-0b1cfb80-de54-11eb-871d-59897ebbf5bd.png">

**Comment by dafeder** [comment]
Just chiming in, that on https://github.com/GetDKAN/dkan we have settled on what I guess I'll call "qs style" serialization for complex GET requests. FWIW this style essentially has native support in PHP and additional support in core Symfony libraries. Struggling with the same issues as everyone, probably going to add some minimal parameters and just add narrative text for now, pointing people to the JSON schema and suggestions for how to serialize.

**Comment by ElectroBuddha** [comment]
Five years later, this feature request is still relevant. Any news on this ?

**Comment by mbraz** [comment]
waiting for "qs style" serialization feature

**Comment by darrelmiller** [comment]
@mbraz You are waiting for the feature in what tooling? Docs, testing, code generators?  Here's how you could describe it in the specification.

```yaml
parameters:
 filterParam:
   in: query
   name: filter
   schema:
     type: object
   style: deepObject
   x-deepObject-style: qs
   explode: true
   description: Options for filtering the results
   required: false
```
I just made that extension up.  That's how we are going to move forward here.

The best way to get a feature into the specification is to propose an extension to a tooling creator, or PR a tool with support.  Adoption of the the extension will be what demonstrates the value and allows us to bring the feature into the specification.  

If you really don't want to go the extension route, you could do this as an alternative.

```yaml
parameters:
 filterParam:
   in: query
   name: filter
   content:
     text/vnd.qs-serialization: {}
   description: Options for filtering the results
   required: false
```

In this case I invented a new text based media type for describing the contents of the filter value.  Media types are not usually used for describing query parameter values, but it is technically allowed by OAS.

**Comment by mbraz** [comment]
Hi, wait in swagger to use in NestJS applications. I'll try
text/vnd.qs-serialization...

Em dom., 5 de mar. de 2023 às 14:34, Darrel ***@***.***>
escreveu:

**Comment by handrews** [comment]
I think the simplest way to allow folks to experiment with different approaches would be to implement #1502 (Support for arbitrary query strings).  That would let folks define entirely custom serialization, perhaps based on media types as @darrelmiller suggests, without having to work around the spec's mechanisms for supporting more common serialization forms.

**Comment by POMXARK** [comment]
https://github.com/abbasudo/laravel-purity

This is extremely necessary! I can't convert to format query parameter

filters[is_publish][$contains]=true

**Comment by dafeder** [comment]
@POMXARK let's keep it respectful, the maintainers don't owe you anything. You're welcome to submit a PR or make and share a third-party package if this is so important to you, we're all just doing our best here.

FWIW I'm not a JS developer but I would still strongly vote for standardizing around "qs style" as it's very intuitive and close enough to other serialization methods to work with in different languages. It should be recognized though that there is really no way to do nested query objects in a GET query string that will be straightforward to document.

**Comment by handrews** [comment]
@dafeder thanks for your comment!

> It should be recognized though that there is really no way to do nested query objects in a GET query string that will be straightforward to document.

Yeah, this is why I'm pushing for #1502 in version 3.2 of the spec.  It's not ideal, but it gets the OpenAPI spec details out of the way and lets folks do their own serialization for the entire query string, including by just setting the whole thing to `content: {application/x-www-form-urlencoded: {...}}` and serializing the way we do `application/x-www-form-urlencoded` bodies.  Which would also allow creating 3rd-party extensions to handle alternate query string formats, which is not possible right now because there is no place in any Object where you could easily put such an extension.

**Comment by DustinCai** [comment]
Any updates on this?

**Comment by handrews** [comment]
@DustinCai see my comment above about 3.2.  We're getting 3.0.4 and 3.1.1 out the door right now, then 3.2 should be a relatively quick release is we want to keep it small (and then do a 3.3 if 3.2 is well-received).  I'd hoped to get 3.2 done by the end of the year, but 3.0.4 and 3.1.1 have taken a bit longer so maybe January or (hopefully) at worst February.

I know that solution (#1502) might not be what you're looking for, but there are too many possible different formats and too many competing demands.  Allowing people to define their own query string formats (media types) for the whole query string will unblock experimentation, particularly if combined with a media type registry per #3771.

**Comment by zachdaniel** [comment]
Is there actually a situation where someone wants the query parameter specified as a deepObject to be keys with url encoded values that are themselves JSON? I feel like I'm misunderstanding the issue, because I've never heard of that as being something that anyone actually wants. `foo[bar][baz]=buz` has been the standard way of encoding "nested" query parameters for pretty much my entire career. At a minimum, it would serve every modern web setup better if that was the one option chosen as opposed to `foo=%7B%22bar%22:%20%22buz%22%7D` which is just wildly strange to me.

The other flaw with that is that it's ambiguous. For instance, our filters support `name="{Fred"` and `name[eq]="Fred"`. But I can't tell if the user is actually typing json into the search box, to look for a thing with the name `{"eq": "Fred"}`, or if they are asking for the thing with the name equal to "Fred". So I can't even add server side logic to properly handle this.

If the default was `foo[bar]=baz`, I could *detect* which format was which, translating that to `foo={"bar": "baz"}` in some preprocessor.

I think I must be understanding because I can't really conceive of how thousands of other people haven't hit on this same issue up to now. Is there a solution I'm not aware of?

**Comment by handrews** [comment]
@zachdaniel it's mostly that `deepObject` is intentionally under-specified, which was intended to allow experimentation and possibly future expansion.  But there isn't really an obvious way to have tools try out different things here.

In any event, please take a look at the just-released-yesterday [3.0.4](https://spec.openapis.org/oas/v3.0.4) ([release notes](https://github.com/OAI/OpenAPI-Specification/releases/tag/3.0.4)) and [3.1.1](https://spec.openapis.org/oas/v3.1.1) ([release notes](https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.1)) releases, as they both contain a substantial rewrite and dramatic expansion (with multiple new appendices added to the spec) of the parameter and form serialization rules.

**Comment by zachdaniel** [comment]
Will check it out. I also may have been wrong about the form style. I think I still haven't figured out how to get the format I'm looking for yet.

**Comment by zachdaniel** [comment]
I still don't see something hinting at what, to me, would be the most common option. What is the way to tell it that a query parameter with nested schemas should be encoded as `foo[bar][baz]=buz`. I cannot find it. Or are you saying there is just no standard way to do this and individual tools (swagger UI) in this case will just do whatever they want?

**Comment by zachdaniel** [comment]
I get it from the spec perspective, but ultimately as someone maintaining a framework offering automatic swagger UI generation, this one singular sticking point issue is causing users to decide not to use swagger UI at all. It has (for them) a sort of all-or-nothing effect because they want to expose this to API consumers. I may be able to figure something out along the lines of detecting a request coming through swagger UI's "try it out" feature (maybe some way to put in a header to all of those requests for example), which would allow me to then do special query parameter decoding.

I think I'm primarily taking umbrage with SwaggerUI at this point, not the spec. I can't imagine a world where anyone wants a query parameter to be URI encoded JSON. I get it that it's unspecified, but thats a weird choice. Not much left to do here on this issue, and looking forward to a resolution to #1502 as potentially opening a path here.

**Comment by handrews** [comment]
> I think I'm primarily taking umbrage with SwaggerUI at this point, not the spec.

Yeah we can't help you with that, I'm afraid!

> I can't imagine a world where anyone wants a query parameter to be URI encoded JSON.

Oh, I've seen it.  I can't say I'd _recommend_ it, but I've definitely seen it.  I've also seen JSON documents embedded in JSON strings.  People do some astonishing things...

> EDIT: nevermind, that won't work either because I still can't tell the difference between the value being a string of valid JSON and a query parameter encoded by swagger UI 😢

I'm not quite sure where the JSON complaint comes in, though.  Is SwaggerUI treating `deepObject` as an instruction to encode like JSON?  It is definitely not supposed to do that.

**Comment by zachdaniel** [comment]
As far as I can tell, it's when I have a deepObject where a value is an `object`. It ends up encoded as `field="{"some": "json"}"`

**Comment by handrews** [comment]
@zachdaniel huh.  Well, the spec does not define behavior for that case, so on the one hand, we can't say that's _wrong_.  But also if they've chosen to do that, that's thier own decision and they don't need anything from us if they want to make a different decision.  As far as the spec is concerned,  the behavior of `deepObject` with properties that are themselves objects or arrays is undefined, so anything that a tool does is entirely under their own control (and unlikely to be portable).

**Comment by handrews** [comment]
@zachdaniel although if this is an `application/x-www-form-urlencoded` request body, they might be applying the Encoding Object's logic for serializing JSON objects as embedded JSON?  That's the only thing I can think of that might be going on (and, despite having extensively researched and mostly re-written that entire part of the spec over the last few months, i can't recall exactly how those requirements would interact).

**Comment by spuxx-dev** [comment]
I struggle to understand what the current status of this is and what solution is considered idiomatic. It appears to me that this thread kinda points at Swagger UI, while Swagger UI kinda points at the spec: https://github.com/swagger-api/swagger-ui/issues/7734

Query parameter syntax like `filter?[foo][bar]=xyz` is and has been quite common, so I feel there should be a straight-forward way to represent those in OpenAPI (and SwaggerUI).

https://github.com/swagger-api/swagger-ui/issues/7734

**Comment by spuxx-dev** [comment]
Thank you for the in-depth explanation @handrews 😊

**Comment by nnnikolay** [comment]
I'm not sure how my solution is correct, however, I needed to fix it, and here is what I did for that.

in the `apiinner.mustache`

I changed that

```
                {{^isPrimitiveType}}
                {{^isEnumRef}}
                {{^isEnum}}
                for (const [key, value] of Object.entries({{paramName}})) {
                    localVarQueryParameter[key] = value;
                }
                {{/isEnum}}
                {{/isEnumRef}}
```

for this
```
                  {{^isPrimitiveType}}
                    {{^isEnumRef}}
                    {{^isEnum}}
                    // deepObject serialization for nested params (e.g. filter[field]=…)
                    for (const [key, value] of Object.entries({{paramName}})) {
                        if (Array.isArray(value)) {
                            value.forEach(item => {
                                localVarQueryParameter[`{{baseName}}[${key}]`] = item;
                            });
                        } else {
                            localVarQueryParameter[`{{baseName}}[${key}]`] = value;
                        }
                    }
                    {{/isEnum}}
                    {{/isEnumRef}}
```

and then I give it up to the axios + qs to handle the parameters properly

```js
import axios, { AxiosInstance, InternalAxiosRequestConfig, AxiosHeaders } from 'axios';
import qs from 'qs';

const sharedAxios: AxiosInstance = axios.create({
  paramsSerializer: (params) => qs.stringify(params, { encodeValuesOnly: true, arrayFormat: 'repeat' }),
});
```

which I also imported via modification of the baseApi.mustache
```
import sharedAxios from 'my-shared-axios-instance';
const globalAxios = sharedAxios;
```

It may not be a great solution because it may block me from the upgrade, but at least it works for me as expected.

**Comment by handrews** [comment]
We ended up adding `in: querystring` to 3.2, so this can be closed as "as fixed as it's going to get."
